lunatrace/bsl/backend/src/github/projects.ts

```diff
@@ -11,42 +11,111 @@
  * limitations under the License.
  *
  */
-import { graphql, GraphqlResponseError } from '@octokit/graphql';
+import { graphql } from '@octokit/graphql';
 import { Configuration, Identity, V0alpha2ApiFactory } from '@ory/kratos-client';
+import deepmerge from 'deepmerge';
 import express from 'express';
 
-import { callHasura } from '../hasura-calls/baseHasuraClient';
-
-const operation = `
-    mutation GetUserAccessToken($user_id: uuid!) {
-        insert_builds_one(object: {project_id: $project_id, s3_url: $s3_url}) {
-          id
-        }
-    }
-`;
-
-export async function getUserGithubAccessToken(variables: { user_id: string }): Promise<string> {
-  const data = await callHasura(operation, 'InsertBuild', variables);
-  if (!data.insert_builds_one) {
-    console.error('Hasura response missing fields ', data);
-    throw new Error('Failed to download manifest for processing');
-  }
-  return '';
-}
+import { hasura } from '../hasura-api';
+import {
+  AuthorizedUserOrganizationsQuery,
+  Identities_Constraint,
+  Identities_Obj_Rel_Insert_Input,
+  Organization_User_Constraint,
+  Organization_User_Insert_Input,
+  Organizations_Constraint,
+  Organizations_Insert_Input,
+  Organizations_Obj_Rel_Insert_Input,
+} from '../hasura-api/generated';
+
+import { GetUserOrganizationsQuery } from './generated';
+import { pullDataForInstallation } from './installation-populate';
+
+import { generateGithubGraphqlClient } from './index';
 
 export const githubApiRouter = express.Router();
 
-export interface KratosIdentityConfig {
-  providers: KratosIdentityProvider[];
-}
+githubApiRouter.get('/webhook/github/install', async (req, res) => {
+  const installationIdQueryParam = req.query.installation_id;
+  const setupActionQueryParam = req.query.setup_action;
+
+  const error = req.query.error;
+  const error_description = req.query.error_description;
+  const error_uri = req.query.error_uri;
+
+  if (error) {
+    res.status(401).send({
+      error: error,
+      description: error_description,
+      uri: error_uri,
+    });
+    return;
+  }
 
-export interface KratosIdentityProvider {
-  initial_id_token: string;
-  subject: string;
-  provider: string;
-  initial_access_token: string;
-  initial_refresh_token: string;
-}
+  if (!installationIdQueryParam || typeof installationIdQueryParam !== 'string') {
+    res.status(401).send({
+      error: 'installation_id not provided in query params',
+    });
+    return;
+  }
+
+  const installationId = parseInt(installationIdQueryParam, 10);
+
+  const installationData = await pullDataForInstallation(installationId);
+
+  const {
+    data: { repositories },
+  } = installationData;
+
+  type OrganizationInputLookup = Record<string, Organizations_Insert_Input>;
```

Comment:
Nit: This could likely benefit from being pulled out into another file. The `reduce` below feels pretty self-contained and potentially useful elsewhere in the app. Also, it clutters this "controller" function a bit too much because it's effectively business logic and not just request/response value mapping.